ros_ws/src/autonomous/src/wall_following.cpp

```diff
@@ -0,0 +1,336 @@
+	float theta = 95;
+
+	// two distances a and b
+	// with b being the distance at angle 0 and a being the distance at angle theta
+	// the lidar has a range of 270°, the algorithm use 180°
+	// so the the angle 0° for the algorith starts at angle 45° for the lidar
+	float a = getRange(lidar,theta);
+	float b = getRange(lidar,45);
+
+	ROS_INFO_STREAM("right a: " << a);
+	ROS_INFO_STREAM("right b: " << b);
+
+	// transform the difference between the angles at a and b into radians
+	float swing = (fabs(theta - 45) * M_PI) / 180;
+
+	// alpha is the orientation of the car with respect to the wall
+	// and a_b is the distance to the wall
+	// a_c is the distance between the current position of the car and the predicted position
+	// we use the car_length as a_c for now PLEASE CHANGE IN THE FUTURE
+	// c_d is the distance to the wall at the predicted position
+	float alpha = atan((a*cos(swing)-b)/(a*sin(swing)));
+	float a_b = b*cos(alpha);
+	float a_c = 0.5;
+	float c_d = a_b + a_c*sin(alpha);
+
+	ROS_INFO_STREAM("right_alpha: " << alpha);
+	ROS_INFO_STREAM("right_a_b: " << a_b);
+	ROS_INFO_STREAM("right_c_d: " << c_d);
+
+	// values used for PID control, given be the F1Tenth tutorial
+	float kp = 14;
+	float kd = 0.09;
+
+	// if we want to stay 0.5 meter away from the wall
+	// then the error is 0.5-CD
+	g_right_error = 0.5 - c_d;
+
+	ROS_INFO_STREAM("right error: " << g_right_error);
+
+	// scale error so the car can react fast enough
+	g_right_error = 5*g_right_error;
+
+	ROS_INFO_STREAM("scaled_right error: " << g_right_error);
+
+	// correction according to the F1Tenth tutorial
+	float correction = kp*g_right_error + kd*(g_right_prev_error - g_right_error);
+
+	//std::cout << "correction:" << correction << std::endl;
+	ROS_INFO_STREAM("right correction: " << correction);
+
+	g_right_prev_error = g_right_error;
+
+	// use correction to increment or decrement steering angle
+	g_right_corrected_angle = (correction * M_PI) / 180;
+
+	// boundary check to avoid extrem angle values
+	if(g_right_corrected_angle > 20)
+	{
+		g_right_corrected_angle = 20;
+	}
+	if(g_right_corrected_angle < -20)
+	{
+		g_right_corrected_angle = -20;
+	}
+
+	float g_right_velocity = 4 - fabs(g_right_corrected_angle);
+
+	if(g_right_velocity > 4)
+	{
+		g_right_velocity = 4;
+	}
+	if(g_right_velocity < 0.3)
+	{
+		g_right_velocity = 0.3;
+	}
+
+	return {g_right_corrected_angle, g_right_velocity};
+}
+
+/**
+ * @brief Using the lidar scans this method calculates the necesarry angle and velocity
+ * for the car to stay parallel to the left wall and then returns them both in that order.
+ * 
+ * @param lidar 
+ * @return std::array<float, 2> 
+ */
+std::array<float, 2> followLeftWall(const sensor_msgs::LaserScan::ConstPtr& lidar)
+{
+	// calculations done according to the F1Tenth tutorial
+	// http://f1tenth.org/lab_instructions/t6.pdf
+
+	// some angle theta between 0 and 70 which you can choose by yourself
+	float theta = 175;
```

Comment:
Why is the angle theta 175? Also does not the comment say it should be between 0 and 70? What does 0 and 70 mean?